Sources/SwiftSMTP/String-Wrap.swift

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright Stephen Beitzel 2021
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+import Foundation
+
+/// This extension is intended to encapsulate the wrap/unwrap
+/// functionality described at https://datatracker.ietf.org/doc/html/rfc5322#section-2.2.3
+
+extension String {
+
+    /// Given the desired line length, split this string into
+    /// a sequence of lines of nearly that length, then join them
+    /// back together with the sequence `\r\n `
+    public func wrap(_ to: Int = 78) -> String {
```

Comment:
To be consistent with the terminology in RFC5322 2.2.3, we should probably call these functions "foldHeader" and "unfoldHeader".

Also, I think it would be nice to have the parameter be an enum, something like:

 think at this point it would be nice to have a more dedicated API for this purpose.  Perhaps something like:

```swift
enum FoldLineLength {
   /// RFC5322 maximum header length
   case maximum
   /// RFC5322 recommended header length
   case recommended
   case custom(Int)

   var value: Int {
      switch self {
      case .maximum: return 998
      case .recommended: return 78
      case .custom(let length): return length
   }
}
```

The API would then look like this: 

```swift

func foldHeader(_ lineLength: FoldLineLength = .recommended) -> String
```

My intention here is mainly for clarity (i.e. why 78?) and partially to provide for future customizability.

Reply:
In addition, I think these methods should be marked `internal` unless there is any significant need for applications to have access to these methods?

Reply:
That's an interesting question. I came to this project because I was looking for a way to build a play-by-email game server entirely in Swift. This project provides half the capability - sending email - but the other half, receiving email, doesn't have a Swift implementation that I could find. A POP3 client *may* have need for such a feature; I honestly have not yet started work on creating such a client, so I don't know.

Reply:
I think it's safe to keep it internal until such a need arises.  I don't like to pollute the namespace especially for common Classes.  It can be jarring for people to see these things as side effects when they're simply importing a library.

Reply:
No, you're absolutely right, we don't want to make any ugly surprises. Honestly, I'm starting to wonder if an extension on String is even the right way to go, or if it would be better to make a separate function that accepts a String and a desired length. Something like:

```
/// Error type thrown by fold/unfold functions
enum FoldError: Error { case maxLengthExceeded }

/// Fold a header, if necessary, to respect the provided line length
///
/// If the folded header still contains a line that is longer than the
/// maximum length specified, then throw an error.
func fold(_ header: String, at lineLength: FoldLineLength = .recommended,
      maxLength: FoldLineLength = .maximum) throws -> String

// Unfold a header, converting the fold token to a single whitespace.
func unfold(_ header: String) -> String
```

Reply:
I don't think it should be a global function.  As an internal extension on String I think it's fine as long as it's clear that it's for the header... either by naming it `foldHeader()` or `fold(header: String,...)`.  Alternatively we could perhaps have a HeaderValue class/struct that has a "values" array as well as a string conversion for use in a header.  It feels a bit heavy for what we're doing, but it might be a cleaner approach as it would fully encapsulate the different ways in which we can use the header values while still using the dictionary.  It still feels a little heavy for what we're doing though.

Reply:
I've moved the functionality inside Mail and kept it private, so I think that addresses the whole namespace/heaviness problem.